bindings/go/blob/direct/serial/blob.go

```diff
@@ -0,0 +1,102 @@
+package serial
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+)
+
+// Blob is a blob that hands out [io.ReadCloser] instances ("leases") that serialize access to a single
+// underlying [io.Reader]. You can call ReadCloser at any time, also concurrently; the *first* [io.Reader.Read]
+// on that lease will block until the previous lease is closed (or until the
+// context is canceled, whatever comes first).
+//
+// IMPORTANT: The underlying reader is shared and stateful. Each lease continues
+// reading from wherever the previous lease stopped, if it does not support io.Seeker.
+//
+// If the stateful reader does support io.Seeker, acquiring a lease will reset the reader position to io.SeekStart.
+// If you need independent positions and concurrent reads, your source must support io.ReaderAt, at which
+// point it can be used without this serial reader.
+type Blob struct {
+	ctx    context.Context
+	reader io.Reader
+	token  chan struct{} // capacity 1: acts as a binary semaphore
+}
+
+// New creates a Blob within a context. If the context is canceled,
+// any leases being acquired will see ctx.Err().
+func New(ctx context.Context, reader io.Reader) *Blob {
+	s := &Blob{
+		ctx:    ctx,
+		reader: reader,
+		token:  make(chan struct{}, 1),
+	}
+	// Put initial token to indicate "available".
+	s.token <- struct{}{}
+	return s
+}
+
+// ReadCloser returns a leased reader immediately. The first Read() on that
+// lease will try to acquire the token (blocking until available or until ctx is canceled).
+func (s *Blob) ReadCloser() (io.ReadCloser, error) {
+	l := &lease{
+		parent: s,
+		reader: s.reader,
+	}
+
+	// acquired is true if the token was acquired.
+	var acquired atomic.Bool
+
+	// Lazily acquire on the first Read, at most once, and cache the error.
+	l.acquire = sync.OnceValue(func() error {
+		select {
+		case <-l.parent.token:
+			acquired.Store(true)
+		case <-s.ctx.Done():
+			return s.ctx.Err()
+		}
+		// If underlying reader is a seeker, reset to start for this lease.
+		// Otherwise, the reader is stateful and we can't reset. Then we do best effort.
+		// and leave it up to the caller to handle errors like
+		if seeker, ok := l.reader.(io.Seeker); ok {
+			_, err := seeker.Seek(0, io.SeekStart)
+			return err
+		}
+		return nil
+	})
+	l.close = func() error {
+		if !acquired.Load() {
+			// Release the token if we did acquire it before.
+			l.parent.token <- struct{}{}
+		}
```

Comment:
Am I stupid or is this comment wrong? :D

Reply:
Yeah, but aren't you supposed to `if acquired.Load()` instead of `if !acquired.Load()`? Since `// acquired is true if the token was acquired.`

Reply:
you are completely right >.<. thanks for catching this!